src/main/java/org/jitsi/videobridge/cc/AdaptiveTrackProjectionContext.java

```diff
@@ -0,0 +1,72 @@
+/*
+ * Copyright @ 2019 8x8, Inc
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.jitsi.videobridge.cc;
+
+import org.jitsi.impl.neomedia.rtp.*;
+import org.jitsi.service.neomedia.*;
+
+/**
+ * @author George Politis
+ */
+public interface AdaptiveTrackProjectionContext
+{
+    /**
+     * An empty {@link RawPacket} array that is used as a return value and
+     * indicates that the input packet needs to be dropped.
+     */
+    RawPacket[] DROP_PACKET_ARR = new RawPacket[0];
+
+    /**
+     * An empty {@link RawPacket} array that is used as a return value when no
+     * packets need to be piggy-backed.
+     */
+    RawPacket[] EMPTY_PACKET_ARR = new RawPacket[0];
+
+    /**
+     * Determines whether a packet needs to be accepted or not.
+     *
+     * @param rtpPacket the RTP packet to determine whether to project or not.
+     * @param targetIndex the target quality index
+     * @return true if the packet is accepted, false otherwise.
+     */
+    boolean accept(RawPacket rtpPacket, int targetIndex);
+
+    /**
+     * @return true if this instance needs a keyframe, false otherwise.
+     */
+    boolean needsKeyframe();
+
+    /**
+     * Rewrites the RTP packet that is specified as an argument.
```

Comment:
> can you add more detail? rewrites in what way(s)? i assume something like `rewrite the timestamp, sequence number and ssrc to keep the rtp stream consistent` ?

👍 



> also, i'm a bit confused by the return value: why is this named around rewriting but then also returns whether or not it should be accepted? it seems like those 2 bits of logic could be separated

I know it's not pretty and it looks confusing; here I need a way to return an error value, just in case something goes wrong, for example, because of an exotic case we've accepted a packet, so we've created a VP8 frame projection, but when it's time to rewrite the packet we can't find a frame projection. Can you think of a better way to deal with this sort of handling?

Reply:
sounds like a place where an exception might be appropriate?  i think it would just be caught and the error logged by the layer above.

Reply:
What about https://github.com/jitsi/jitsi-videobridge/pull/735/commits/7b387bcd611e81f1b98622bf959491bb16bcd25d?